RavenDB-19470 Handle the leaks of RqlQueries on the SYNC Postgre message

poissoncorp opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19470/Suddenly-220GB-of-scratch-buffers-created-taking-all-space

### Additional description

The scratch buffers were created by multiple hanging read transactions, that were created inside the PowerBI integration `RqlQuery `class, which executes queries using a `QueryOperationContext`.
While calling `RqlQuery.Dispose()`, `QueryOperationContext` should be disposed and thus transactions should be closed but we've leaked `RqlQuery` inside the `Sync()` method of the `PgTransaction` class.
As a result of the leak query contexts weren't disposed and transactions remained open.

After the fix, we don't leak RqlQueries anymore and dispose them properly on the SYNC messages.
In addition, I've added an 'emergency' RqlQuery _destructor_ that'll log such errors in the future and handle the leaks.


### Type of change

- Bug fix

### How risky is the change?

- Low 

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

karmeli87 commented on src/Raven.Server/Integrations/PostgreSQL/RqlQuery.cs:
```diff
@@ -335,6 +349,7 @@ protected virtual void HandleSpecialColumnsIfNeeded(string columnName, Blittable
 
         public override void Dispose()
         {
+            GC.SuppressFinalize(this);
```
How do we dispose the context if `IsNamedStatement` is `true` ?

arekpalinski replied:
Answered in https://github.com/ravendb/ravendb/pull/15226#issuecomment-1308791819

poissoncorp committed:
RavenDB-19470 Dispose RqlQuery inside the Sync() correctly

poissoncorp committed:
RavenDB-19488 Add 'emergency' destructor of the RqlQuery to prevent unrecognized leaks

poissoncorp commented:
With @arekpalinski, we've discovered that the issue directly relates to the[ Azure Data Factory PR](https://issues.hibernatingrhinos.com/issue/RavenDB-18716/PostgreSQL-connector-is-not-working-with-Azure-Data-Factory) .
Previously, we've added support for caching named statements - storing the `PgQuery` inside `NamedStatements` dict. It forced us to omit the query context disposing in the `Sync()`. But there's a better way to handle both situations. We shouldn't create an instance of the RqlQuery just on the constructor, but inside the RunRqlQuery - just where it is _needed_.

I am testing the fix currently, so I'll convert this PR to draft.

poissoncorp committed:
RavenDB-19470 Changed the timing of disposing the query context to the Execute() 'finally' - we materialize the query context only for the time of its' need

poissoncorp commented:
I've changed the logic of creating and disposing a query context. Now, we dispose it in the `Execute()` `finally`. We instantiate the query context only for the time of its need. I've tested it on many scenarios including remote PowerBI data gateway connection.

@karmeli87, your question was right.
We were aware of disposing the PgQuery on SYNC messages that will cause the dispose of QueryContext (waiting to be used in the next message). Thats why we included that `isNamedStatement` check in the `Dispose()`. The "DEALLOCATE" message has used to point out that now we can dispose the query context because we don't need this prepared statement anymore.
 
Now, following the changes, we don't need to check if the query is a prepared statement, because we aren't aware of that. QueryContext is instantiated on the RunRqlQuery() and disposed just after the execution. (In addition, I've deleted the `IsNamedStatement` property, because now it provides useless information). 

By going this way, we'll handle both prepared statements support and RqlQuery leaks.
I'll write a prepared statements test with Npgsql

arekpalinski commented on src/Raven.Server/Integrations/PostgreSQL/RqlQuery.cs:
```diff
@@ -17,21 +17,32 @@
 
+        ~RqlQuery()
+        {
+            Logger.Operations($"Query '{this.QueryString}' wasn't disposed properly.\n" +
```
Please add `if (Logger.IsOperationsEnabled)`

arekpalinski replied:
Instead of `\n` it's better to use `Environment.NewLine` so it will be formatted correctly on Windows too

arekpalinski commented:
@poissoncorp The PR should be for `v5.3` branch

poissoncorp committed:
RavenDB-19470 Support for named statements

poissoncorp committed:
RavenDB-19470 Named Statements Test

poissoncorp committed:
RavenDB-19470 Enhance the RqlQuery destructor

poissoncorp commented:
Rebased on v5.3 https://github.com/ravendb/ravendb/pull/15261
